source/interprocedural_analyses/taint/taintConfiguration.ml

```diff
@@ -1517,6 +1545,49 @@ let validate ({ Heap.sources; sinks; transforms; features; _ } as configuration)
     |> Result.combine_errors_unit
     |> Result.map_error ~f:List.concat
   in
+  let compare_locations previous_location location =
+    match previous_location, location with
+    | Some previous_location, Some location ->
+        JsonAst.LocationWithPath.equal previous_location location
+    | None, None -> true
+    | _ -> false
+  in
+  let ensure_list_unique_with_location ~get_key ~get_error ~get_location elements =
+    let table = Hashtbl.create (module String) in
+    let seen_errors = Hash_set.create (module Error) in
+    let ensure_unique element =
+      let key = get_key element in
+      let location = get_location element in
+      match Hashtbl.find table key with
+      | None ->
+          let _ = Hashtbl.add table ~key ~data:location in
+          Result.Ok ()
+      | Some previous_location when compare_locations previous_location location ->
+          (* Some generated rules can have the same code, compare the base defintions *)
+          Result.Ok ()
+      | Some previous_location ->
+          let error =
+            match location with
+            | Some location ->
+                Error.create_with_location
+                  ~path:location.JsonAst.LocationWithPath.path
+                  ~location:location.JsonAst.LocationWithPath.location
+                  ~kind:(get_error element previous_location)
+            | None ->
+                { Error.path = None; location = None; kind = get_error element previous_location }
+          in
+          (* Since there might be two generated rules with the same code, when we encounter another,
+             if we don't compare, we might throw in duplicate errors.*)
+          if Hash_set.mem seen_errors error then
+            Result.Ok ()
+          else (
+            Hash_set.add seen_errors error;
+            Result.Error [error])
```

Comment:
Could we go back to using `List.sort_and_dedup`? I would avoid hash sets in ocaml and use the "functional" approach. It's fine if we need to implement `compare` on `Error.t`.

Reply:
`List.dedup_and_sort` takes a tiny bit more of performance but okay!